File: src/com/train/algorithm/twoPointers/SubarraysWithKDifferentIntegers.cpp
//
// Created by 罗斯 on 2/9/2022.
//

#include <vector>

using namespace std;
class Solution {
public:
    int subarraysWithKDistinct(vector<int>& nums, int k) {
        auto subarrays = [&nums](int k){
            vector<int> count(nums.size() + 1);
            int res = 0;
            int i = 0;
            for(int j = 0;j < nums.size();++j){
                if(count[nums[j]]++ == 0)
                    --k;
                while (k < 0)
                    if(--count[nums[i++]] == 0)
                        ++k;
                res += j - i + 1;
            }
            return res;
        };
        return subarrays(k) - subarrays(k - 1);
    }
};

File: src/com/train/algorithm/twoPointers/implementInPython/SubarraysWithKDifferentIntegers.py
class Solution(object):
    def atMostKDistinct(self, nums, k):
        hashmap = {}
        subarrays = 0
        start, end = 0, 0
        while end < len(nums):
            if nums[end] in hashmap:
                hashmap[nums[end]] += 1
            else:
                hashmap[nums[end]] = 1
            while len(hashmap) > k:
                hashmap[nums[start]] -= 1
                if hashmap[nums[start]] == 0:
                    del hashmap[nums[start]]
                start += 1
            subarrays += end - start + 1
            end += 1
        return subarrays
    def subarraysWithKDistinct(self, nums, k):
        return self.atMostKDistinct(nums, k) - self.atMostKDistinct(nums, k - 1)
